### ISSUE TITLE
uniqueName for Users

### ISSUE THREAD

**Issue by slavizh** [microsoftgraph/msgraph-bicep-types#135]
**Is your feature request related to a problem? Please describe.**
Users can be members of groups. Due to that would be good if you can add them to groups using existing syntax. Currently that is not possible due to users not being part of the resources available in Bicep Graph.

**Describe the solution you'd like**
With this I would like not specifically to be able to deploy users with Bicep Graph but to at least being able to reference them by unique name. As uniqueName is something that can be updated by Graph API and it is something that most Graph resources in Bicep have would makes things easier if you can use existing syntax for users before being able to deploy the users via Bicep. I would assume just adding uniqueName to user resources it will be a lot easier than just making it available for creation via Bicep.

**Additional context**

**Comment by dkershaw10** [comment]
@eketo-msft we should take a look at this.  This makes sense to me.  I wonder if we should do the same for devices, whiel we are at it.  @slavizh already has a legitimate ask for adding `uniqueName` property to groups too - #128

**Comment by slavizh** [comment]
Additionally to add scenario to this - there are many Azure resources that allow configuring (via Bicep) Microsoft Entra users, groups or service principals for authenticating to the service's data plane.

**Comment by eketo-msft** [comment]
@slavizh, would userPrincipalName work for your scenarios?  It's not an alternate key (officially) and it can change, but it is unique and there is reluctance to add additional "name" properties to the User resource.

**Comment by slavizh** [comment]
@eketo-msft yes it should be sufficient.

**Comment by eketo-msft** [comment]
Quick update:  We added support for UPN as an alternative key on the User resource in MS Graph.  Next up we're working on Bicep type support so it can be used with the existing keyword.  That is likely to land in early December due to lockdowns for the US elections and Thanksgiving holidays.

**Comment by slavizh** [comment]
@eketo-msft awesome! that will make referencing users even more easier.

**Comment by eketo-msft** [comment]
@alex-frankel, exactly.  It will allow reading an existing user by UPN and then referencing the user for operations, like managing membership, which require `objectId` (or `id` in MS Graph).

**Comment by alex-frankel** [comment]
Amazing! You will make a lot of people happy with this one!

**Comment by Gijsreyn** [comment]
I don't know if this feature was already implemented, but I would like the ability to fetch it in reverse, so in this case:

```bicep
resource user 'Microsoft.Graph/users@v1.0' existing = {
  objectId: deployer().objectId
}

output upn string = user.userprincipalName
```

This allows me to use it in the `api-management` module as such:

```bicep
module apiManagement 'br/public:avm/res/api-management/service:0.6.0' = {
    name: 'apiManagement'
    params: {
        name: 'apim${resourceGroup().name}'
        publisherEmail: upn
        publisherName: upn
        sku: 'Basic'
        managedIdentities: {systemAssigned: true}
    }
}
```

**Comment by dkershaw10** [comment]
@Gijsreyn The new (read-only) Bicep type should be available later this month.
It won't feature the reverse-lookup as far as I know (as the main requirement here was to expose a unique friendly name for users which has a better experience in declarative config-as-code files). 

Maybe a better bet would be to have the `deployer()` function also return UPN? We'll also look into whether we can do a reverse lookup.

Another thing to note, is that while the email address is often set to the same value as UPN, that is not always the case, so you might be setting the publisherEmail incorrectly. Better to use the user's `mail` property.

**Comment by Gijsreyn** [comment]
Hey @dkershaw10. You are right, I want the email property in the above use case. It would be pretty cool to see this functionality being added to Bicep in someway. Thanks for the quick response.

**Comment by dkershaw10** [comment]
Once this feature ships, you'll be able to look up a user by UPN and get various key properties for the user, including their mail address. You can what the `users` Bicep type will look like here: https://github.com/microsoftgraph/msgraph-bicep-types/blob/main/generated/microsoftgraph/microsoft.graph/v1.0/0.1.9-preview/types.md#resource-microsoftgraphusersv10-readonly

@alex-frankel would you be open to updating the deployer() function to return UPN as well?

**Comment by Gijsreyn** [comment]
Thanks Dan, looking forward to it.

**Comment by alex-frankel** [comment]
Tagging @levimatheri who implemented the `deployer()` function. If memory serves, it was not cheap to return UPN, which is why it got scoped out, but let's see.

**Comment by eketo-msft** [comment]
This is working as of Bicep v0.33.

`
extension 'br:mcr.microsoft.com/bicep/extensions/microsoftgraph/v1.0:0.1.9-preview' as graph

resource user 'Microsoft.Graph/users@v1.0' existing = {
  userPrincipalName: 'person@contoso.com'
}

output id string = user.id
`

**Comment by levimatheri** [comment]
> Tagging [@levimatheri](https://github.com/levimatheri) who implemented the `deployer()` function. If memory serves, it was not cheap to return UPN, which is why it got scoped out, but let's see.

Update: We revisited this, and found out we should be able to add `userPrincipalName` to the `deployer()` function cheaply.

**Comment by dkershaw10** [comment]
> This is working as of Bicep v0.33.
> 
> extension 'br:mcr.microsoft.com/bicep/extensions/microsoftgraph/v1.0:0.1.9-preview' as graph
> 
> resource user 'Microsoft.Graph/users@v1.0' existing = { userPrincipalName: '[person@contoso.com](mailto:person@contoso.com)' }
> 
> output id string = user.id

Sample demonstrating this is here: https://github.com/microsoftgraph/msgraph-bicep-types/tree/main/quickstart-templates/security-group-add-user-members

**NOTE** that it requires use of dynamic types and the 0.1.9 Graph Bicep types release.
